### PR TITLE
Use `raise XslxValueError` and other error handling improvements

### LINKED ISSUE #233: xlsx2csv does not raise exceptions for fatal errors, which makes it hard to use as a library.
xlsx2csv does not raise exceptions for fatal errors, which makes it hard to use as a library.

There are several instances where `eprint()` followed by sys.exit(1) or `return 1` are used.
https://github.com/dilshod/xlsx2csv/blob/de96b659b7ce531d08246b912fa654a4ded62629/xlsx2csv.py#L131-L132

To have it kind of working as a library, I have to overwrite `eprint` with my own function that raises an exception:
https://github.com/pola-rs/polars/pull/3567/commits/65464c225762d7c92c7e7f3aa8920974a3b3329a#diff-5f0fd72b20838297a2173cfaafc9b6b1348b05eb43f1a47ee3d6f4c8cf73abe6R1087-R1092

It would be great if it would be properly handled in xlsx2csv so polars can support reading Excel files without this hack: https://github.com/pola-rs/polars/pull/3567

### PR DESCRIPTION
**Pull request by cnpryer**
fixes #233 by 

- [x] replacing `eprint` calls with direct `XlsxValueError` raises
- [x] use `sys.exit` directly

### CODE REVIEW

**Comment by cnpryer** [comment]
I'll fix the indentation in a bit

**Comment by cnpryer** [comment]
what if we did a little more and also replaced `eprint` with `raise`es where `eprint` was called and also where `sys.stderr.write` is called?
https://github.com/dilshod/xlsx2csv/blob/ee4d40cf543903a44083a92c97a1912af719ff0e/xlsx2csv.py#L1118

File: xlsx2csv.py [hunk 1]
```diff
@@ -1206,8 +1195,7 @@ def main():
             xlsx2csv.convert(outfile, sheetid)
     except XlsxException:
         _, e, _ = sys.exc_info()
-        sys.stderr.write(str(e) + "\n")
```
Comment (ghuls) [inline xlsx2csv.py:1209]: Here you should not raise an exception, but print the string to stderr

File: xlsx2csv.py [hunk 2]
```diff
@@ -148,7 +148,6 @@ class SheetNotFoundException(XlsxException):
 class OutFileAlreadyExistsException(XlsxException):
     pass
 
-
```
Comment (ghuls) [inline xlsx2csv.py:151]: class XlsxValueError(XlsxException):
    pass

File: xlsx2csv.py [hunk 3]
```diff
@@ -285,17 +284,15 @@ def _convert(self, sheet_index, outfile):
...
             closefile = True
         try:
             writer = csv.writer(outfile, quoting=self.options['quoting'], delimiter=self.options['delimiter'],
                                 lineterminator=self.options['lineterminator'])
 
             sheets_filtered = list(filter(lambda s: s['index'] == sheet_index, self.workbook.sheets))
             if len(sheets_filtered) == 0:
-                eprint("Sheet with index %i not found or can't be handled" % sheet_index)
-                return 1
+                raise XlsxException("Sheet with index %i not found or can't be handled" % sheet_index)
```
Comment (ghuls) [inline xlsx2csv.py:295]: Use XlsxValueError

File: xlsx2csv.py [hunk 4]
```diff
@@ -808,8 +805,7 @@ def handleCharData(self, data):
 
                 # get format type
                 if not format_str:
-                    eprint("unknown format %s at %d" % (format_str, xfs_numfmt))
-                    return
+                    raise XlsxException("unknown format %s at %d" % (format_str, xfs_numfmt))
```
Comment (ghuls) [inline xlsx2csv.py:808]: Use XlsxValueError

File: xlsx2csv.py [hunk 5]
```diff
@@ -870,9 +866,7 @@ def handleCharData(self, data):
                         self.data = ("%f" % (float(self.data))).rstrip('0').rstrip('.')
 
                 except (ValueError, OverflowError):  # this catch must be removed, it's hiding potential problems
-                    eprint("Error: potential invalid date format.")
-                    # invalid date format
-                    pass
+                    raise XlsxException("Error: potential invalid date format.")
```
Comment (ghuls) [inline xlsx2csv.py:869]: XlsxValueError

File: xlsx2csv.py [hunk 6]
```diff
@@ -1115,8 +1109,7 @@ def main():
         (options, args) = parser.parse_args()
         if len(args) < 1:
             parser.print_usage()
-            sys.stderr.write("error: too few arguments" + os.linesep)
-            sys.exit(1)
+            raise XlsxException("error: too few arguments" + os.linesep)
```
Comment (ghuls) [inline xlsx2csv.py:1112]: In main, don't trow exceptions but print the error message to stderr

File: xlsx2csv.py [hunk 7]
```diff
@@ -1129,8 +1122,7 @@ def main():
     elif options.delimiter[0] == 'x':
         options.delimiter = chr(int(options.delimiter[1:]))
     else:
-        sys.stderr.write("error: invalid delimiter\n")
-        sys.exit(1)
+        raise XlsxException("error: invalid delimiter\n")
```
Comment (ghuls) [inline xlsx2csv.py:1125]: Idem.

File: xlsx2csv.py [hunk 8]
```diff
@@ -1141,8 +1133,7 @@ def main():
     elif options.quoting == 'all':
         options.quoting = csv.QUOTE_ALL
     else:
-        sys.stderr.write("error: invalid quoting\n")
-        sys.exit(1)
+        raise XlsxException("error: invalid quoting\n")
```
Comment (ghuls) [inline xlsx2csv.py:1136]: Idem.

File: xlsx2csv.py [hunk 9]
```diff
@@ -1153,8 +1144,7 @@ def main():
     elif options.lineterminator == '\\r\\n':
         options.lineterminator = '\r\n'
     else:
-        sys.stderr.write("error: invalid line terminator\n")
-        sys.exit(1)
+        raise XlsxException("error: invalid line terminator\n")
```
Comment (ghuls) [inline xlsx2csv.py:1147]: Idem.

File: xlsx2csv.py [hunk 10]
```diff
@@ -1165,8 +1155,7 @@ def main():
     elif options.sheetdelimiter[0] == 'x':
         options.sheetdelimiter = chr(int(options.sheetdelimiter[1:]))
     else:
-        sys.stderr.write("error: invalid sheet delimiter\n")
-        sys.exit(1)
+        raise XlsxException("error: invalid sheet delimiter\n")
```
Comment (ghuls) [inline xlsx2csv.py:1158]: Idem.

**Comment by cnpryer** [comment]
@ghuls! You're the best. Thanks for the pointers.

Now that I'm curious, is it more ideal to use `sys.stderr.write` + a `sys.exit` or should this just throw a `sys.exit`?

https://docs.python.org/3/library/sys.html#sys.exit

File: xlsx2csv.py (hunk 6 above)
Comment (cnpryer) [reply xlsx2csv.py:1112]: Thanks! Curious of the motivation here if you don't mind sharing 😊

Comment (ghuls) [reply xlsx2csv.py:1112]: main is used when you run xlsx2csv as command line application. So a nicely written message in a proper formatted way looks the best in the terminal. When xls2csv is imported as a library, main is not called (only convert, which raises exceptions which can be handled).

**Comment by ghuls** [comment]
> @ghuls! You're the best. Thanks for the pointers.
> 
> Now that I'm curious, is it more ideal to use `sys.stderr.write` + a `sys.exit` or should this just throw a `sys.exit`?
> 
> https://docs.python.org/3/library/sys.html#sys.exit

Seems like sys.exit is viable for writing the error message.

**Comment by cnpryer** [comment]
What about this guy under the `main`
https://github.com/dilshod/xlsx2csv/blob/ee4d40cf543903a44083a92c97a1912af719ff0e/xlsx2csv.py#L1205

**Comment by ghuls** [comment]
It should be hanlded by: https://github.com/dilshod/xlsx2csv/blob/ee4d40cf543903a44083a92c97a1912af719ff0e/xlsx2csv.py#L1205 , but could also be handled explicitly. https://github.com/dilshod/xlsx2csv/blob/ee4d40cf543903a44083a92c97a1912af719ff0e/xlsx2csv.py#L1205


This should also raise a InvalidXlsxFileException exception: ttps://github.com/dilshod/xlsx2csv/blob/ee4d40cf543903a44083a92c97a1912af719ff0e/xlsx2csv.py#L1039

File: xlsx2csv.py [hunk 11]
```diff
@@ -285,17 +289,15 @@ def _convert(self, sheet_index, outfile):
             elif sys.version_info[0] == 3:
                 outfile = open(outfile, 'w+', encoding=self.options['outputencoding'], newline="")
             else:
-                sys.stderr.write("error: version of your python is not supported: " + str(sys.version_info) + "\n")
-                sys.exit(1)
+                raise XlsxException("error: version of your python is not supported: " + str(sys.version_info) + "\n")
```
Comment (cnpryer) [inline xlsx2csv.py:292]: Not sure if it really matters, but I could do something with this too.

File: xlsx2csv.py (hunk 6 above)
Comment (cnpryer) [reply xlsx2csv.py]: makes sense 👍 thanks for explaining

File: xlsx2csv.py [hunk 12]
```diff
@@ -1202,12 +1196,11 @@ def main():
             if options.sheetname:
                 sheetid = xlsx2csv.getSheetIdByName(options.sheetname)
                 if not sheetid:
-                    raise XlsxException("Sheet '%s' not found" % options.sheetname)
+                    sys.exit("Sheet '%s' not found" % options.sheetname)
```
Comment (cnpryer) [inline xlsx2csv.py:1199]: If handled explicitly means something like `SheetNotFoundException` I could do that. I just thought that ince we're in `main` I'd use `sys.exit`

Comment (ghuls) [reply xlsx2csv.py:1199]: I mean, it can use sys.exit there.

**Pull request merged**